README.md
# HealthHacks

Website/changelog.txt
Version: 4.7.1
  - Fixed pricing table borders issue

Version: 4.7.0
  - Updated Bootstrap to version 5.1.3
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.6.1
  - Fixed the hero section, which was not displaying properly in mobile landscape mode

Version: 4.6.0
  - Updated Bootstrap to version 5.1.2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.5.0
  - Fixed slider issue in testimonials and portfolio details sections

Version: 4.4.0
  - Updated Bootstrap to version 5.1.1
  - Updated all outdated third party vendor libraries to their latest versions
  - Improved and updated dev version gulp scripts

Version: 4.3.0
  - Updated Bootstrap to version 5.0.1
  - Updated all outdated third party vendor libraries to their latest versions
  - Fixed navigation links focus color

Version: 4.2.0
  - Updated Bootstrap to version 5.0.0 Final
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.1.0
  - Updated Bootstrap to version 5.0.0-beta3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to V3.1

Version: 4.0.1
  - Updated Bootstrap to version 5.0.0-beta2
  - Updated all outdated third party vendor libraries to their latest versions

Version: 4.0.0
  - The template does not require jQuery anymore
  - Removed jQuery and all the jQuery plugins
  - The assets/js/main.js was rewritten completely with vanilla Javascript. No more jQuery code
  - Restructured the dev version for better development experience
  - Updated the PHP Email Form to V3.0 - No jQuery dependency. Added attachment support

Version: 3.0.0
  - Initial release with Bootstrap v5.0 Beta 1

Version: 2.2.0
  - Updated Bootstrap to version 4.5.3
  - Updated all outdated third party vendor libraries to their latest versions
  - Updated the PHP Email Form to v2.3
  - Other small fixes and improvements

Version: 2.1.0
  - Updated Bootstrap to version 4.5.0
  - Updated the PHP Email Form library to version 2.0 with reCaptcha support
  - AAded inner-page.html page template
  - Added smooth scroll on page load with hash links in the url
  - Added navigation active state on scroll
  - Replaced the WOW animation on scroll library with AOS
  - Updated all outdated third party vendor libraries to their latest versions
  - Other small fixes and improvements
  
Version: 2.0.0
  - The template was rebuilt from scratch with the latest Bootstrap version (4.4.1)
  - Added SMPTP support for the contact form script (Pro)
  - Added NodeJS NPM Development version (Pro unlimited & Membership members)
  
Version: 1.0.0
- Initial Release


Website/backend/main.py
from flask import Flask, redirect, url_for, render_template, request, flash
# from firebase_admin import credentials, auth, firestore
import pyrebase
# import firebase_admin
import qrcode
import json

BASE_URL = 'http://127.0.0.1:5000'

# Connect to Firebase Firestore
# cred = credentials.Certificate('firebase-admin.json')
# firebase_admin.initialize_app(cred)
# db = firestore.client()

# Connect to Firebase Realtime Database
config = {
  "apiKey": "AIzaSyBGRn-c9Cl6kbROuWu44w5besGDNxsDM5I",
  "authDomain": "retro-c0350.firebaseapp.com",
  "projectId": "retro-c0350",
  "databaseURL": "https://retro-c0350-default-rtdb.firebaseio.com/",
  "storageBucket": "retro-c0350.appspot.com",
  "messagingSenderId": "112454151211",
  "appId": "1:112454151211:web:abd98d22947b2b37ba0e6e"
}

firebase = pyrebase.initialize_app(config)
auth = firebase.auth()
db = firebase.database()


app = Flask(__name__)

# Views
@app.route('/')
def index():
    return render_template('index.html')


@app.route('/signup', methods = ["POST", "GET"])
def signup():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        name = result["name"]
        date_of_birth = result["date_of_birth"]
        if email is None or password is None or name is None or date_of_birth is None:
            return {'message': 'Error missing information'}, 400
        try:
            auth.create_user_with_email_and_password(email, password)
            print('success')
            user = auth.sign_in_with_email_and_password(email, password)
            print(user['localId'])
            # id = user['localId']
            url = f'{BASE_URL}/portal/{id}'
            print('url created')

            # Firebase Realtime Database
            data = {
                "id": id,
                "name": name,
                "email": email,
                "date_of_birth": date_of_birth,
                "url": url
            }
            db.child("users").child(id).set(data)
            print('database updated')


            return redirect('portal/{}'.format(user['localId']))
        except:
            print('error creating user')
            return {'message': 'Error creating user'}, 400
    else:
        return render_template('signup.html')


@app.route('/login', methods = ["POST", "GET"])
def login():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        if email is None or password is None:
            return {'message': 'Error missing information'}, 400
        try:
            user = auth.sign_in_with_email_and_password(email, password)
            print(user['localId'])

            return redirect(url_for('portal/{}'.format(user['localId'])))
        except:
            return {'message': 'Error logging in'}, 400
    else:
        return render_template('login.html')


@app.route('/map')
def map():
    return render_template('map.html')

@app.route('/portal/<id>')
def portal(id):
    db = firebase.database()
    user = db.child("users").child(id).get()

    # users = db.child("users").get()



    return render_template('medical_portal.html')


def getQRCode(id):
    img = qrcode.make('{}/portal/{}}'.format(BASE_URL, id))
    type(img)
    return img

if __name__ == '__main__':
    app.run(debug=True)

Website/backend/AFAFA.py
from flask import Flask, redirect, url_for, render_template, request, flash
from firebase_admin import credentials, auth, firestore
import qrcode
import json

BASE_URL = 'http://127.0.0.1:5000'

# Connect to Firebase
cred = credentials.ApplicationDefault()
firebase_admin.initialize_app(cred)
db = firestore.client()

app = Flask(__name__)

# Authentication
# def check_token(f):
#     @wraps(f)
#     def wrap(*args,**kwargs):
#         if not request.headers.get('authorization'):
#             return {'message': 'No token provided'},400
#         try:
#             user = auth.verify_id_token(request.headers['authorization'])
#             request.user = user
#         except:
#             return {'message':'Invalid token provided.'},400
#         return f(*args, **kwargs)
#     return wrap


# @app.route('/api/userinfo')
# @check_token
# def userinfo():
#     return {'data': users}, 200


# Views
@app.route('/')
def index():
    return render_template('index.html')


# @app.route('/signup')
# def signup():
#     return render_template('signup.html')

@app.route('/login')
def login():
    return render_template('login.html')

@app.route('/map')
def map():
    return render_template('map.html')

@app.route('/portal/<username>', methods = ["POST", "GET"])
def portal(username):
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        try:
            user = auth.sign_in_with_email_and_password(email, password)
            global person
            person["is_logged_in"] = True
            person["email"] = user["email"]
            person["uid"] = user["localId"]
            data = db.child("users").get()
            person["name"] = data.val()[person["uid"]]["name"]
            return render_template('medical_portal.html')
        except:
            return redirect('/login')
    else:
        if person["is_logged_in"] == True:
            return render_template('medical_portal.html')
        else:
            return redirect('/login')

    return render_template('medical_portal.html')

@app.route('/register', methods = ["POST", "GET"])
def register():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["pass"]
        name = result["name"]
        try:
            auth.create_user_with_email_and_password(email, password)
            user = auth.sign_in_with_email_and_password(email, password)
            global person
            person["is_logged_in"] = True
            person["email"] = user["email"]
            person["uid"] = user["localId"]
            person["name"] = name
            # EDIT
            data = {"name": name, "email": email}
            db.child("users").child(person["uid"]).set(data)
            return redirect(url_for('medical_portal'))
        except:
            return render_template('register.html')


def getQRCode(username):
    img = qrcode.make(f'{link}/portal/username')
    type(img)
    return img

# @app.route('/medicalInfo', methods = ["POST", "GET"])
# def medicalInfo():




# API
# @app.route('/api/signup')
# def signup():
#     email = request.form.get('email')
#     password = request.form.get('password')
#     if email is None or password is None:
#         return {'message': 'Error missing email or password'},400
#     try:
#         user = auth.create_user(
#                email=email,
#                password=password
#         )

            
        
#         return {'message': f'Successfully created user {user.uid}'},200
#     except:
#         return {'message': 'Error creating user'},400

# @app.route('/api/token')
# def token():
#     email = request.form.get('email')
#     password = request.form.get('password')
#     try:
#         user = pb.auth().sign_in_with_email_and_password(email, password)
#         jwt = user['idToken']
#         return {'token': jwt}, 200
#     except:
#         return {'message': 'There was an error logging in'},400

if __name__ == '__main__':
    app.run(debug=True)

Website/backend/main3.py
from flask import Flask, redirect, url_for, render_template, request, flash
from firebase_admin import credentials, auth
import qrcode
import firebase_admin
import json
import pyrebase


# Connect to firebase
# cred = credentials.Certificate("firebase-admin.json")
# firebase_admin.initialize_app(cred)
# pb = pyrebase.initialize_app(json.load(open('firebase-config.json')))

#initialize firebase
# firebase = pyrebase.initialize_app(config)
# auth = firebase.auth()
# db = firebase.database()

#Initialze person as dictionary
# person = {"is_logged_in": False, "name": "", "email": "", "uid": ""}

app = Flask(__name__)

# Authentication
# def check_token(f):
#     @wraps(f)
#     def wrap(*args,**kwargs):
#         if not request.headers.get('authorization'):
#             return {'message': 'No token provided'},400
#         try:
#             user = auth.verify_id_token(request.headers['authorization'])
#             request.user = user
#         except:
#             return {'message':'Invalid token provided.'},400
#         return f(*args, **kwargs)
#     return wrap


# @app.route('/api/userinfo')
# @check_token
# def userinfo():
#     return {'data': users}, 200


# Views
@app.route('/')
def hello():
    return render_template('index.html')

@app.route('/map')
def map():
    return render_template('map.html')
@app.route('/portal/<username>')
def portal(username):
    # return f'User {escape(username)}'
    return render_template('medical_portal.html')


# API
# @app.route('/api/signup')
# def signup():
#     email = request.form.get('email')
#     password = request.form.get('password')
#     if email is None or password is None:
#         return {'message': 'Error missing email or password'},400
#     try:
#         user = auth.create_user(
#                email=email,
#                password=password
#         )

            
        
#         return {'message': f'Successfully created user {user.uid}'},200
#     except:
#         return {'message': 'Error creating user'},400

# @app.route('/api/token')
# def token():
#     email = request.form.get('email')
#     password = request.form.get('password')
#     try:
#         user = pb.auth().sign_in_with_email_and_password(email, password)
#         jwt = user['idToken']
#         return {'token': jwt}, 200
#     except:
#         return {'message': 'There was an error logging in'},400

if __name__ == '__main__':
    app.run(debug=True)

Website/backend/main4.py
from flask import Flask, redirect, url_for, render_template, request, flash
from datetime import date
import pyrebase
import qrcode
import json
import time
import io
import base64

BASE_URL = 'http://127.0.0.1:5000'

# Connect to Firebase Firestore
# cred = credentials.Certificate('firebase-admin.json')
# firebase_admin.initialize_app(cred)
# db = firestore.client()

# Connect to Firebase Realtime Database
config = {
  "apiKey": "AIzaSyBGRn-c9Cl6kbROuWu44w5besGDNxsDM5I",
  "authDomain": "retro-c0350.firebaseapp.com",
  "projectId": "retro-c0350",
  "databaseURL": "https://retro-c0350-default-rtdb.firebaseio.com/",
  "storageBucket": "retro-c0350.appspot.com",
  "messagingSenderId": "112454151211",
  "appId": "1:112454151211:web:abd98d22947b2b37ba0e6e"
}

firebase = pyrebase.initialize_app(config)
auth = firebase.auth()
db = firebase.database()
storage = firebase.storage()

# path_on_cloud = "images/foo.jpg"
# path_local = "my_image.jpg"
# storage.child(path_on_cloud).put(path_locao)

test_id = 'cEmDKustQ5g7SFrwruRmxlQI7622'

USER = {"idToken": None, "localId": None}

app = Flask(__name__)

# Views
@app.route('/')
def index():
    return render_template('index.html')


@app.route('/signup', methods = ["POST", "GET"])
def signup():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        name = result["name"]
        date_of_birth = result["date_of_birth"]
        sex = result["sex"]

        # date of birth
        today = date.today()

        yr = int(date_of_birth[:4])

        age = today.year - yr
        

        if email is None or password is None or name is None or date_of_birth is None or sex is None:
            return {'message': 'Error missing information'}, 400
        try:
            auth.create_user_with_email_and_password(email, password)
            user = auth.sign_in_with_email_and_password(email, password)
            id = user['localId']
            url = f'{BASE_URL}/portal/{id}'

            # Firebase Realtime Database
            data = {
                "id": id,
                "name": name,
                "email": email,
                "date_of_birth": date_of_birth,
                "url": url,
                "sex": sex,
                "height": "N/A",
                "weight": "N/A",
                "blood_type": "N/A",
                "emergency_phone": "N/A",
                "emergency_email": "N/A",
                "past_conditions": {
                    "initial": "N/A",
                },
                "treatments": {
                    "initial": {
                        "hospital": "N/A",
                        "treatment": "N/A"
                    }
                },
                "blood_tests": "N/A",
                "age": age,
                "files": {
                    "initial": True,
                },
            }
            db.child("users").child(id).set(data)

            USER["idToken"] = user['idToken']
            USER["localId"] = user['localId']

            return redirect(url_for('personal_portal'))
        except:
            print('error creating user')
            return {'message': 'Error creating user'}, 400
    else:
        return render_template('signup.html')


@app.route('/login', methods = ["POST", "GET"])
def login():
    if request.method == "POST":
        result = request.form
        email = result["email"]
        password = result["password"]
        if email is None or password is None:
            return {'message': 'Error missing information'}, 400
        try:
            user = auth.sign_in_with_email_and_password(email, password)
            USER["idToken"] = user['idToken']
            USER["localId"] = user['localId']

            return redirect(url_for('personal_portal'))
        except:
            return {'message': 'Error logging in'}, 400
    else:
        return render_template('login.html')


@app.route('/map')
def map():
    return render_template('map.html')


@app.route('/personal-portal')
def personal_portal():
    db = firebase.database()
    if USER["idToken"] != None and USER["localId"] != None:
        id = USER["localId"]
        user = db.child("users").child(id).get()
        user_data = user.val()
        name = user_data["name"]
        date_of_birth = user_data["date_of_birth"]
        email = user_data["email"]
        url = user_data["url"]
        sex = user_data["sex"]
        height = user_data["height"]
        weight = user_data["weight"]
        blood_type = user_data["blood_type"]
        age = user_data["age"]
        past_conditions = user_data["past_conditions"]
        treatments = user_data["treatments"]
        blood_tests = user_data["blood_tests"]
        emergency_phone = user_data["emergency_phone"]
        emergency_email = user_data["emergency_email"]

 

        data = io.BytesIO()
        img = getQRCode(id)
        img.save(data, "JPEG")
        encoded_img_data = base64.b64encode(data.getvalue())

        return render_template('dashboard/pages/dashboard.html', img_data=encoded_img_data.decode('utf-8'), name=name, email=email, url=url, id=id, age=age, sex=sex, height=height, weight=weight, blood_type=blood_type, past_conditions=past_conditions, treatments=treatments, blood_tests=blood_tests, emergency_phone=emergency_phone, emergency_email=emergency_email)
    else:
        return redirect(url_for('login'))


@app.route('/portal/<id>')
def portal(id):
    db = firebase.database()
    user = db.child("users").child(id).get()
    user_data = user.val()

    id = user_data["id"]
    name = user_data["name"]
    date_of_birth = user_data["date_of_birth"]
    email = user_data["email"]
    url = user_data["url"]
    sex = user_data["sex"]
    height = user_data["height"]
    weight = user_data["weight"]
    blood_type = user_data["blood_type"]
    age = user_data["age"]
    past_conditions = user_data["past_conditions"]
    treatments = user_data["treatments"]
    blood_tests = user_data["blood_tests"]
    emergency_phone = user_data["emergency_phone"]
    emergency_email = user_data["emergency_email"]
    

    data = io.BytesIO()
    img = getQRCode(id)
    img.save(data, "JPEG")
    encoded_img_data = base64.b64encode(data.getvalue())

    # users = db.child("users").get()

    return render_template('dashboard/pages/dashboard.html', img_data=encoded_img_data.decode('utf-8'), name=name, email=email, url=url, id=id, age=age, sex=sex, height=height, weight=weight, blood_type=blood_type, past_conditions=past_conditions, treatments=treatments, blood_tests=blood_tests, emergency_phone=emergency_phone, emergency_email=emergency_email)


@app.route('/add-entry', methods = ["POST", "GET"])
def add_entry():


    if request.method == "POST":
        result = request.form
        height = result["height"]
        weight = result["weight"]
        blood_type = result["blood_type"]
        blood_tests = result["blood_tests"]
        emergency_phone = result["emergency_phone"]
        emergency_email = result["emergency_email"]
        if height is None or weight is None or blood_type is None or blood_tests is None or emergency_phone is None or emergency_email is None:
            return {'message': 'Error missing information'}, 400
        try:
            id = USER['localId']

            # Firebase Realtime Database
            data = {
                "height": height,
                "weight": weight,
                "blood_type": blood_type,
                # "past_conditions": past_conditions,
                # "treatments": treatments,
                "blood_tests": blood_tests,
                "emergency_phone": emergency_phone,
                "emergency_email": emergency_email
            }
            db.child("users").child(id).update(data)


            return redirect(url_for('personal_portal'))
        except:
            print('error updating user')
            return {'message': 'Error updating user'}, 400
    else:
        return render_template('add_entry.html')



@app.route('/add-file', methods = ["POST", "GET"])
def add_file():
    db = firebase.database()
    id = USER["localId"]
    if id == None:
        return redirect(url_for('login'))

    if request.method == "POST":
        file = request.files['file']
        if file.filename == '':
            flash('No selected file')
            return {'message': 'Error missing information'}, 400
        # UPLOAD
        storage_rel_url = f"files/{id}/{time.time()}.pdf"
        storage.child(storage_rel_url).put(file)

        timestamp = str(time.time())[:9]
        
        try:

            # Firebase Realtime Database
            data = {
                timestamp: f"{storage_rel_url}",
            }

            if db.child("users").child(id).child("files").child("initial").get() == None:
                print('hi')
                db.child("users").child(id).child("files").child("initial").remove()
                db.child("users").child(id).child("files").set(data)
            else:
                db.child("users").child(id).child("files").update(data)

            return redirect(url_for('personal_portal'))
        except Exception as e:
            print(e)
            print('error updating user')
            return {'message': 'Error updating user'}, 400
    else:
        return render_template('add_file.html')


@app.route('/view-file')
def view_file():
    db = firebase.database()
    id = USER["localId"]
    if id == None:
        return redirect(url_for('login'))

    user = db.child("users").child(id).get()
    user_data = user.val()

    file_urls = user_data["files"]
    display_urls = []

    for key, url in file_urls.items():
        full_url = storage.child(url).get_url(USER["idToken"])
        display_urls.append(full_url)


    return render_template('view-pdfs.html', urls=display_urls)
    
    




def getQRCode(id):
    img = qrcode.make('{}/portal/{}'.format(BASE_URL, id))
    #type(img)
    return img

if __name__ == '__main__':
    app.run(debug=True)